From: Oever González <software@notengobattery.com>
Subject: [PATCH] mt7621: improve CPU clock
Date: Fri, 21 May 2021 22:20:32 -0600

This patch sets the CPU frequency to the maximum allowed by the PLL.

Signed-off-by: Oever González <software@notengobattery.com>
---
Index: linux-5.15.150/drivers/clk/ralink/clk-mt7621.c
===================================================================
--- linux-5.15.150.orig/drivers/clk/ralink/clk-mt7621.c
+++ linux-5.15.150/drivers/clk/ralink/clk-mt7621.c
@@ -15,12 +15,13 @@
 #include <dt-bindings/clock/mt7621-clk.h>
 
 /* Configuration registers */
-#define SYSC_REG_SYSTEM_CONFIG0         0x10
-#define SYSC_REG_SYSTEM_CONFIG1         0x14
+#define MEMC_REG_CPU_PLL		0x648
+#define MT7621A_TARGET_CLOCK_HZ		1200000000
 #define SYSC_REG_CLKCFG0		0x2c
 #define SYSC_REG_CLKCFG1		0x30
 #define SYSC_REG_CUR_CLK_STS		0x44
-#define MEMC_REG_CPU_PLL		0x648
+#define SYSC_REG_SYSTEM_CONFIG0		0x10
+#define SYSC_REG_SYSTEM_CONFIG1		0x14
 
 #define XTAL_MODE_SEL_MASK		GENMASK(8, 6)
 #define CPU_CLK_SEL_MASK		GENMASK(31, 30)
@@ -250,7 +251,7 @@ static unsigned long mt7621_cpu_recalc_r
 	struct regmap *sysc = clk->priv->sysc;
 	struct regmap *memc = clk->priv->memc;
 	u32 clkcfg, clk_sel, curclk, ffiv, ffrac;
-	u32 pll, prediv, fbdiv;
+	u32 pll, prediv, fbdiv, target_pll, target_fbdiv;
 	unsigned long cpu_clk;
 
 	regmap_read(sysc, SYSC_REG_CLKCFG0, &clkcfg);
@@ -268,6 +269,18 @@ static unsigned long mt7621_cpu_recalc_r
 		regmap_read(memc, MEMC_REG_CPU_PLL, &pll);
 		fbdiv = FIELD_GET(CPU_PLL_FBDIV_MASK, pll);
 		prediv = FIELD_GET(CPU_PLL_PREDIV_MASK, pll);
+		target_fbdiv = (MT7621A_TARGET_CLOCK_HZ * ffiv) / ffrac;
+		target_fbdiv = target_fbdiv << prediv_tbl[prediv];
+		target_fbdiv = (target_fbdiv / xtal_clk) - 1;
+		target_pll = pll & ~FIELD_PREP(CPU_PLL_FBDIV_MASK, fbdiv);
+		target_pll |= FIELD_PREP(CPU_PLL_FBDIV_MASK, target_fbdiv);
+		pr_info("CPU Overclock PLL: 0x%x\n", target_pll); panic("CPU Overclock PLL");
+		regmap_write(memc, MEMC_REG_CPU_PLL, target_pll);
+
+		regmap_read(memc, MEMC_REG_CPU_PLL, &pll);
+		fbdiv = FIELD_GET(CPU_PLL_FBDIV_MASK, pll);
+		prediv = FIELD_GET(CPU_PLL_PREDIV_MASK, pll);
+
 		cpu_clk = ((fbdiv + 1) * xtal_clk) >> prediv_tbl[prediv];
 		break;
 	default:
